kernel/src/callback.rs

```diff
@@ -9,44 +9,122 @@ use crate::process;
+/// serves as a tool for capsules to hold pointers to applications.
+///
+/// Since `AppId` implements `Copy`, having an `AppId` does _not_ ensure that
+/// the process the `AppId` refers to is still valid. The process may have been
+/// removed, terminated, or restarted as a new process. Therefore, all uses of
+/// `AppId` in the kernel must check that the `AppId` is still valid. This check
+/// happens automatically when `.index()` is called, as noted by the return
+/// type: `Option<usize>`. `.index()` will return the index of the process in
+/// the processes array, but if the process no longer exists then `None` is
+/// returned.
+///
+/// Outside of the kernel crate, holders of an `AppId` may want to use `.id()`
+/// to retrieve a simple identifier for the process that can be communicated
+/// over a UART bus or syscall interface. This call is guaranteed to return a
+/// suitable identifier for the `AppId`, but does not check that the
+/// corresponding application still exists.
+///
+/// This type also provides capsules an interface for interacting with processes
+/// since they otherwise would have no reference to a `ProcessType`. Very limited
+/// operations are available through this interface since capsules should not
+/// need to know the details of any given process. However, certain information
+/// makes certain capsules possible to implement. For example, capsules can use
+/// the `get_editable_flash_range()` function so they can safely allow an app
+/// to modify its own flash.
 #[derive(Clone, Copy)]
 pub struct AppId {
+    /// Reference to the main kernel struct. This is needed for checking on
+    /// certain properties of the referred app (like its editable bounds), but
+    /// also for checking that the index is valid.
     crate kernel: &'static Kernel,
-    idx: usize,
+
+    /// The index in the kernel.PROCESSES[] array where this app's state is
+    /// stored. This makes for fast lookup of the process and helps with
+    /// implementing IPC.
+    ///
+    /// This value is crate visible to enable optimizations in sched.rs. Other
+    /// users should call `.index()` instead.
+    crate index: usize,
+
+    /// The unique identifier for this process. This can be used to refer to the
+    /// process in situations where a single number is required, for instance
+    /// when referring to specific applications across the syscall interface.
+    ///
+    /// The combination of (index, identifier) is used to check if the app this
+    /// `AppId` refers to is still valid. If the stored identifier in the
+    /// process at the given index does not match the value saved here, then the
+    /// process moved or otherwise ended, and this `AppId` is no longer valid.
+    identifier: usize,
 }
 
 impl PartialEq for AppId {
     fn eq(&self, other: &AppId) -> bool {
-        self.idx == other.idx
+        self.index == other.index && self.identifier == other.identifier
```

Comment:
I think in the current implementation it is probably more correct to compare both fields. The reason is that using an appid with a valid identifier but the wrong index will not work, but using one with both fields correct will. It would be a little strange to have `AppIdA == AppIdB`, but then using `AppIdA` fails while `AppIdB` works.

At one point I did implement the logic for an AppId to automatically update its index, but since processes cannot currently move that seemed like unnecessary complexity right now.

Reply:
Hm, I'm not sure I'm convinced. The opposite argument here is that the `.id()` method is provided to
> retrieve a simple identifier for the process that can be communicated over a UART bus or syscall interface. This call is guaranteed to return a suitable identifier for the `AppId`

So now we have a situation where it's possible that `AppIdA.id() == AppIdB.id()` but `AppIdA != AppIdB`.

I feel like the identifier _is_ the identifier of an app, and the index is just an internal optimization that saves an otherwise expensive lookup.

What situation would arise where the identifiers match but the indexes don't?

Reply:
> What situation would arise where the identifiers match but the indexes don't?

The kernel moves the application to a different index in the processes array. But, there isn't any functionality to do this, so it's a case that doesn't matter.

The fact that processes cannot move slots in the array (since there is currently no code that will do that) can support either implementation of `eq`. However, since just using `identifier` would be required if there ever was a reason to move processes around in the array, that implementation makes more sense. I will update.